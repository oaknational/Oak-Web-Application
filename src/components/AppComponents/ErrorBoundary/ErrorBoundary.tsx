import React, { ErrorInfo, FC, useMemo } from "react";
import Bugsnag from "@bugsnag/js";
import * as Sentry from "@sentry/nextjs";

import ErrorView from "@/components/AppComponents/ErrorView";
import { bugsnagInitialised } from "@/browser-lib/bugsnag/useBugsnag";

const ClientErrorView: FC = () => {
  return (
    <ErrorView headerVariant="client-error" footerVariant="client-error" />
  );
};

type FallbackComponentProps = {
  error: Error;
  info: ErrorInfo;
  clearError: () => void;
};

const FallbackComponent: FC<FallbackComponentProps> = () => {
  return <ClientErrorView />;
};

export type ErrorBoundaryProps = {
  children?: React.ReactNode;
};

/**
 * Wrapper around the respective error boundaries for both Bugsnag and Sentry.
 * This will catch any uncaught errors and show the user {@link ClientErrorView},
 * whilst also reporting the error to the respective error reporting service
 * (if consent has been given).
 *
 * We can safely use both the Bugsnag and Sentry error boundaries without
 * risking they'll send error reports when the user has not consented to the
 * relevant cookie policy. This is because the error boundaries will only
 * send error reports if the service has been initialised, which only happens
 * if the user has consented.
 */
const ErrorBoundary: FC<ErrorBoundaryProps> = (props) => {
  const isBugsnagInitialised = bugsnagInitialised();
  const BugsnagErrorBoundary = useMemo(() => {
    if (isBugsnagInitialised) {
      return Bugsnag.getPlugin("react")?.createErrorBoundary(React);
    }
  }, [isBugsnagInitialised]);

  return (
    <Sentry.ErrorBoundary fallback={<ClientErrorView />}>
      {BugsnagErrorBoundary ? (
        <BugsnagErrorBoundary
          FallbackComponent={FallbackComponent}
          {...props}
        />
      ) : (
        props.children
      )}
    </Sentry.ErrorBoundary>
  );
};

export default ErrorBoundary;
