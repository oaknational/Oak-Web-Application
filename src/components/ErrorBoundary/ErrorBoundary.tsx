import React, { ErrorInfo, FC } from "react";
import Bugsnag from "@bugsnag/js";

import Layout from "../Layout";
import { initialiseBugsnag } from "../../common-lib/error-handler";

type FallbackComponentProps = {
  error: Error;
  info: ErrorInfo;
  clearError: () => void;
};
const FallbackComponent: FC<FallbackComponentProps> = (props) => {
  console.log(props);

  return (
    <Layout>
      <h1>Error</h1>
      <p>
        Sorry, what a mess. Looks like something's gone wrong, the team has been
        notified!
      </p>
    </Layout>
  );
};

// This should happen once per app load.
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
if (!Bugsnag._client) {
  initialiseBugsnag();
}

const BugsnagErrorBoundary =
  Bugsnag.getPlugin("react")?.createErrorBoundary(React);

const ErrorBoundary: FC = ({ children }) => {
  if (!BugsnagErrorBoundary) {
    return null;
  }

  return (
    <BugsnagErrorBoundary FallbackComponent={FallbackComponent}>
      {children}
    </BugsnagErrorBoundary>
  );
};

export default ErrorBoundary;
