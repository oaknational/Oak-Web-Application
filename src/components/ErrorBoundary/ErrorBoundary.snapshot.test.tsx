import { render } from "@testing-library/react";

import ErrorBoundary from "./ErrorBoundary";

it("renders <ErrorBoundary> unchanged", () => {
  const { container } = render(<ErrorBoundary>Contents</ErrorBoundary>);
  expect(container).toMatchSnapshot();
});
