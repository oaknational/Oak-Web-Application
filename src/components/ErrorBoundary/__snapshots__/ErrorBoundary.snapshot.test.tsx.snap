// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <ErrorBoundary> unchanged 1`] = `
<div>
  Contents
</div>
`;
